x-pack/plugins/security/server/lib/authentication/providers/kerberos.ts

```diff
@@ -0,0 +1,280 @@
+    request.headers.authorization = `Bearer ${accessToken}`;
+
+    try {
+      const user = await this.options.client.callWithRequest(request, 'shield.authenticate');
+
+      this.debug('User has been authenticated with new access token');
+
+      return AuthenticationResult.succeeded(user, { accessToken });
+    } catch (err) {
+      this.debug(`Failed to authenticate request via access token: ${err.message}`);
+
+      // Restore `Authorization` header we've just set. We can end up here only if newly generated
+      // access token was rejected by Elasticsearch for some reason and it doesn't make any sense to
+      // keep it in the request object since it can confuse other consumers of the request down the
+      // line (e.g. in the next authentication provider).
+      request.headers.authorization = originalAuthorizationHeader;
+
+      return AuthenticationResult.failed(err);
+    }
+  }
+
+  /**
+   * Tries to extract access token from state and adds it to the request before it's
+   * forwarded to Elasticsearch backend.
+   * @param request Request instance.
+   * @param state State value previously stored by the provider.
+   */
+  private async authenticateViaState(request: Legacy.Request, { accessToken }: ProviderState) {
+    this.debug('Trying to authenticate via state.');
+
+    if (!accessToken) {
+      this.debug('Access token is not found in state.');
+      return AuthenticationResult.notHandled();
+    }
+
+    request.headers.authorization = `Bearer ${accessToken}`;
+
+    try {
+      const user = await this.options.client.callWithRequest(request, 'shield.authenticate');
+
+      this.debug('Request has been authenticated via state.');
+      return AuthenticationResult.succeeded(user);
+    } catch (err) {
+      this.debug(`Failed to authenticate request via state: ${err.message}`);
+
+      // Reset `Authorization` header we've just set. We know for sure that it hasn't been defined before,
+      // otherwise it would have been used or completely rejected by the `authenticateViaHeader`.
+      // We can't just set `authorization` to `undefined` or `null`, we should remove this property
+      // entirely, otherwise `authorization` header without value will cause `callWithRequest` to fail if
+      // it's called with this request once again down the line (e.g. in the next authentication provider).
+      delete request.headers.authorization;
+
+      return AuthenticationResult.failed(err);
+    }
+  }
+
+  /**
+   * Tries to query Elasticsearch and see if we can rely on SPNEGO to authenticate user.
+   * @param request Request instance.
+   * @param [state] Optional state object associated with the provider.
+   */
+  private async authenticateViaSPNEGO(request: Legacy.Request, state?: ProviderState | null) {
+    this.debug('Trying to authenticate request via SPNEGO.');
+
+    // If client can't handle redirect response, basically if it's an AJAX request, we shouldn't
```

Comment:
__note:__ I'm hesitating about this one, I'd rather "fix" logged_out page and allow AJAX request to start session as long as SPNEGO token is valid... But didn't think it through, maybe I'm missing some cases when it'd be a problem?

Reply:
Yeah, that was my concern as well, feels like something that will be happening frequently.